src/identity/hcs/did/hcs-did.ts

```diff
@@ -0,0 +1,337 @@
+import {HederaDid} from "../../hedera-did";
+import {DidSyntax} from "../../did-syntax";
+import {FileId, PrivateKey, PublicKey, TopicId} from "@hashgraph/sdk";
+import {Hashing} from "../../../utils/hashing";
+import {DidDocumentBase} from "../../did-document-base";
+import {HcsDidRootKey} from "./hcs-did-root-key";
+
+/**
+ * Hedera Decentralized Identifier for Hedera DID Method specification based on HCS.
+ */
+export class HcsDid implements HederaDid {
+    public static DID_METHOD = DidSyntax.Method.HEDERA_HCS;
+    private static DID_PARAMETER_VALUE_PARTS = 2;
+
+    private didTopicId: TopicId;
+    private addressBookFileId: FileId;
+    private network: string;
+    private idString: string;
+    private did: string;
+    private didRootKey: PublicKey;
+    private privateDidRootKey: PrivateKey;
+
+    /**
+     * Creates a DID instance.
+     *
+     * @param network           The Hedera DID network.
+     * @param didRootKey        The public key from which DID is derived.
+     * @param addressBookFileId The appent's address book {@link FileId}
+     * @param didTopicId        The appnet's DID topic ID.
+     */
+    constructor(network: string, didRootKey: PublicKey, addressBookFileId: FileId, didTopicId?: TopicId);
+    /**
+     * Creates a DID instance with private DID root key.
+     *
+     * @param network           The Hedera DID network.
+     * @param privateDidRootKey The private DID root key.
+     * @param addressBookFileId The appent's address book {@link FileId}
+     * @param didTopicId        The appnet's DID topic ID.
+     */
+    constructor(network: string, privateDidRootKey: PrivateKey, addressBookFileId: FileId, didTopicId?: TopicId);
+    /**
+     * Creates a DID instance without topic ID specification.
+     *
+     * @param network           The Hedera DID network.
+     * @param didRootKey        The public key from which DID is derived.
+     * @param addressBookFileId The appent's address book {@link FileId}
+     */
+    constructor(network: string, didRootKey: PublicKey, addressBookFileId: FileId)
+    /**
+     * Creates a DID instance.
+     *
+     * @param network           The Hedera DID network.
+     * @param idString          The id-string of a DID.
+     * @param addressBookFileId The appent's address book {@link FileId}
+     * @param didTopicId        The appnet's DID topic ID.
+     */
+    constructor(network: string, idString: string, addressBookFileId: FileId, didTopicId?: TopicId);
+    constructor(...args: any[]) {
```

Comment:
What is the advantage of having this constructor and having all these if blocks, vs. just using the other constructors?

Reply:
javascript (and typescript) does not have method or constructor overload, so there is only one constructor block. But we can declare multiple constructors for typings.